File: sujeong/kakao2021/광고 삽입.swift
import Foundation

var timeline:[Double] = []
func solution(_ play_time:String, _ adv_time:String, _ logs:[String]) -> String {
    let play_time_sec = getSecondsOf(play_time)
    
    timeline = Array(repeating: 0, count:play_time_sec+1)
    
    let adv_time_sec = getSecondsOf(adv_time)
    
    for log in logs {
        let log_components = log.components(separatedBy:"-")
        let start = getSecondsOf(log_components[0])
        let end = getSecondsOf(log_components[1])
        timeline[start] += 1
        timeline[end] -= 1
    }
    
    for sec in 1...play_time_sec {
        timeline[sec] += timeline[sec-1]
    }
    
    var sumOfPlay = timeline[0..<adv_time_sec].reduce(0,+)
    
    var maxSumOfPlay:Double = sumOfPlay
    
    var startSecond = 0
    
    if play_time_sec > adv_time_sec {
        for s in 1...(play_time_sec-adv_time_sec) {
            let end = s+adv_time_sec-1
            sumOfPlay = sumOfPlay - timeline[s-1] + timeline[end]

            if maxSumOfPlay < sumOfPlay {
                maxSumOfPlay = sumOfPlay
                startSecond = s
            }
        }
    }
    
    let result:String = getTimeStringOf(startSecond)
    return result
}

func getSecondsOf(_ time:String) -> Int{
    let times = time.components(separatedBy:":")
    let hour = Int(times[0])!*3600
    let min = Int(times[1])!*60
    let sec = Int(times[2])!
    
    return hour + min + sec
}

func getTimeStringOf(_ seconds:Int) -> String {
    var result:String = ""
    
    var second = seconds
    
    let hour:Int = second / 3600
    second %= 3600
    let min:Int = second / 60
    second %= 60
    
    result += hour < 10 ? "0\(hour):" : "\(hour):"
    result += min < 10 ? "0\(min):" : "\(min):"
    result += second < 10 ? "0\(second)" : "\(second)"

    return result
}

File: chaewon/kakao2021/광고삽입.py
def get_time(time):
    hour, minute, second = map(int, time.split(":"))
    return hour * 3600 + minute * 60 + second


def get_string(time):
    hour = str(time // 3600).zfill(2)
    minute = str(time % 3600 // 60).zfill(2)
    second = str(time % 3600 % 60).zfill(2)
    return hour + ":" + minute + ":" + second


def solution(play_time, adv_time, logs):
    play_time = get_time(play_time)
    adv_time = get_time(adv_time)
    all_time = [0 for i in range(play_time + 1)]

    for log in logs:
        start, end = log.split('-')
        start = get_time(start)
        end = get_time(end)
        all_time[start] += 1
        all_time[end] -= 1

    for i in range(1, len(all_time)):
        all_time[i] = all_time[i] + all_time[i - 1]
    for i in range(1, len(all_time)):
        all_time[i] = all_time[i] + all_time[i - 1]

    most_view, max_time = 0, 0
    for i in range(adv_time - 1, play_time):
        if i >= adv_time:
            temp = all_time[i] - all_time[i - adv_time]
            if most_view < temp:
                most_view = temp
                max_time = i - adv_time + 1
        else:
            if most_view < all_time[i]:
                most_view = all_time[i]
                max_time = i - adv_time + 1

    return get_string(max_time)